docs/design/better_coordination_multi_operator.md

```diff
@@ -0,0 +1,129 @@
+- `AddReadyForRestart`: Adds the process group ID to a set of process groups that are ready to be restarted.
+- `GetPendingForRemoval`: Gets the process group IDs for all process groups that are marked for removal.
+- `GetPendingForExclusion`: Gets the process group IDs for all process groups that should be excluded.
+- `GetPendingForInclusion`: Gets the process group IDs for all the process groups that should be included.
+- `GetPendingForRestart`: Gets the process group IDs for all the process groups that should be restarted.
+- `GetReadyForExclusion`: Gets the process group IDs for all the process groups that are ready to be excluded.
+- `GetReadyForInlusion`: Gets the process group IDs for all the process groups that are ready to be included.
+- `GetReadyForRestart`: Gets the process group IDs fir akk tge process groups that are ready to be restarted.
+
+
+The information will be stored in FDB itself, so it will be available for all operators and will get the same benefits from the transaction guarantees.
+The keys will be prefixed with the `client.cluster.GetLockPrefix()` (default `\xff\xff/org.foundationdb.kubernetes-operator`) and the according path, e.g. `readyForExclusion`.
+The value will be empty, except for the `readyForExclusion` and `readyForInclusion` case, in those cases the value will be the process group address(es).
+
+The following sub-reconciler will be modified:
+
+- `updateStatus`: Will add process groups that are marked for removal to the key range and removes them from the key range if they are not being present anymore in the `FoundationDBCluster` status and the process group prefix matches. This sub-reconciler will also remove process groups from the `readyForExclusion`, `readyForInclusion` and `readyForRestart` (and the same for pending).
+- `excludeProcesses`: Will add process groups that are pending for exclusions and add process groups that are ready for exclusion.
+- `changeCoordinators`: Will read the `pendingForRemoval` set and will not take any process group from this set as a candidate for the new coordinators.
+- `removeProcessGroups`: Will add process groups that can be included to `readyForInclusion`.
+
+We will add a new setting under the `FoundationDBClusterAutomationOptions` with the name `synronizationMode`.
+The default will be `local`, which will represent the current state and the new mechanism can be enabled by setting the value to `global`.
+Those changes will add some additional load to the cluster, but the additional load should be fairly limited compared to the customer load on those clusters and the request pattern will be mostly reading the data.
+
+Pseudo-Code changes for the `excludeProcesses` sub-reconciler:
+
+```go
+func (e excludeProcesses) reconcile(ctx context.Context, r *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster, status *fdbv1beta2.FoundationDBStatus, logger logr.Logger) *requeue {
```

Comment:
My interpretation of this flow is that the reconciler is checking for pending/ready exclusions that another operator instance has reported, but doesn't necessarily wait for any kind of delay or convergence before performing the known exclusions itself. Is that accurate? Does that mean that we still have the possibility for each instance of the operator to perform its own exclusions without seeing the exclusions from the other instances?

Reply:
It might be good to get a more detailed flow of how the exclusion states for different operator instances will interleave, and what kinds of things we expect to see as blockers or synchronization points between `PendingForExclusion` and `ReadyForExclusion`.